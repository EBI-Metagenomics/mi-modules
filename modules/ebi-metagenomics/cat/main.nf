process CAT {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/cat:5.3--hdfd78af_0':
        'biocontainers/cat:5.3--hdfd78af_0' }"

    input:
    tuple val(meta), path(contigs_fasta)
    path cat_db
    path taxonomy_db

    output:
    tuple val(meta), path('*_summary.txt'), emit: summary
    tuple val(meta), path('*_contig2classification_official_names.tsv'), emit: contigs_classification
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    echo "Analysing contigs"
    CAT contigs \
        -n ${task.cpus} \
        -c ${contigs_fasta} \
        -d ${cat_db} \
        -t ${taxonomy_db} \
        --out_prefix ${prefix}

    echo "Adding taxonomy names"
    CAT add_names \
        -i ${prefix}.contig2classification.txt \
        -o ${prefix}_contig2classification_official_names.tsv \
        -t ${taxonomy_db} --only_official

    echo "Summarizing output"
    CAT summarise -c ${contigs_fasta} \
        -i ${prefix}_contig2classification_official_names.tsv \
        -o ${prefix}_summary.txt

    ############################################################
    # Storage optimization                                     #
    # Getting rid of files that are not needed after it runs   #
    ############################################################
    rm -f "${prefix}.alignment.diamond"
    rm -f "${prefix}.predicted_proteins.faa"
    rm -f "${prefix}.predicted_proteins.gff"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        CAT: \$(CAT --version | sed "s/CAT v//; s/(.*//")
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_contig2classification_official_names.tsv ${prefix}_summary.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        CAT: \$(CAT --version | sed "s/CAT v//; s/(.*//")
    END_VERSIONS
    """
}
