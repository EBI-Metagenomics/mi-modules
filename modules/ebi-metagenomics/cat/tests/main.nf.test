
nextflow_process {

    name "Test Process CAT"
    script "../main.nf"
    process "CAT"

    tag "modules"
    tag "modules_nfcore"
    tag "cat"

    test("bins-fasta - cat taxonomy") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file("${moduleDir}/tests/fixtures/contigs.fasta", checkIfExists: true)
                ]
                input[1] = file("${moduleDir}/tests/fixtures/cat_db/db", checkIfExists: true)
                input[2] = file("${moduleDir}/tests/fixtures/cat_db/tax", checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("bins-fasta - cat taxonomy - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file("${moduleDir}/tests/fixtures/contigs.fasta", checkIfExists: true)
                ]
                input[1] = file("${moduleDir}/tests/fixtures/cat_db/db", checkIfExists: true)
                input[2] = file("${moduleDir}/tests/fixtures/cat_db/tax", checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
