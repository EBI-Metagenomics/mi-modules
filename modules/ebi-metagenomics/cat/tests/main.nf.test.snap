{
    "bins-fasta - cat taxonomy - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_summary.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_contig2classification_official_names.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,0f2032bf504c399364490799f5962802"
                ],
                "contigs_classification": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_contig2classification_official_names.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "summary": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_summary.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0f2032bf504c399364490799f5962802"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-16T14:53:22.855283"
    },
    "bins-fasta - cat taxonomy": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_summary.txt:md5,f7eed0869f97e803c22a1a4f586aa7c9"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_contig2classification_official_names.tsv:md5,1efb34f5b59ea561115ce262277d1a37"
                    ]
                ],
                "2": [
                    "versions.yml:md5,0f2032bf504c399364490799f5962802"
                ],
                "contigs_classification": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_contig2classification_official_names.tsv:md5,1efb34f5b59ea561115ce262277d1a37"
                    ]
                ],
                "summary": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_summary.txt:md5,f7eed0869f97e803c22a1a4f586aa7c9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0f2032bf504c399364490799f5962802"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-16T14:53:17.194335"
    }
}